Show Identifier Step, pre-register identifiers

kshepherd opened the pull request:
_Attribution: Donated by The Library Code with the support of Technische Universität Hamburg and Technische Informationsbibliothek_

## References
_Add references/links to any related issues or PRs. These may include:_
* Related to Angular UI implementation at: DSpace/dspace-angular#1740
* Related to REST Contract PR DSpace/RestContract#199

## Description
This feature introduces the ability to pre-register identifiers (eg. handle, DOI) and display them in a section during submission and workflow. This is useful to allow users to reference their permanent URIs in the document they upload, other metadata, and so on. This feature is compatible with Logical Item Filtering.

This offers the ability to:

1. Configure the registration of identifiers  at the time of workspace item creation, and dynamic mint/delete of DOIs as in-submission item metadata changes during workspace or workflow.
2. Display a step showing Handle, DOI, and "Other" identifiers that have already been registered for the in-progress item, according to the identifier service.
3. Administrators can register DOIs for individual items in the item status page

List of changes in this PR:
* A new submission step `ShowIdentifiersStep` is created. This is a simple, passive step with no patch operations. It is defined in item submission XML but not included by default in any submission processes.
* A test in `WorkspaceItemServiceImpl.create()` is added, so that if `identifiers.submission.register` is true (default: false), the `IdentifierService.register()` method is called for the underlying item.
* `DOIConsumer.consume()` also checks for this configuration if the item is in workspace or workflow. If true, a DOI is minted or deleted depending on current DOI status and Item Filter test result.
* A new `SubmissionShowIdentifiersRestIT` integration test class is added with a simple test to check for populated DOI and handle identifiers in an "identifiers" section for a new item.
* A new `dspace/config/modules/identifiers.cfg` configuration file with the property to allow pre-registration is added and included in dspace.cfg
* A new (admin only) item status feature Register DOI will queue a pending DOI for registration, or mint and queue a new DOI if there is no pending DOI from submission.

To test this PR, configure the DOI consumer as normal, and set the `identifiers.submission.register` configuration property to true. Include the "identifiers" step in a submission process and test a submission to it.

If testing DOI registration with logical item filtering, note what happens if you use a simple filter that looks at the title value, and you make changes to the title form field, then save (or use autosave), switching the item between "true" and "false" evaluations of the filter. The DOI should be minted or deleted as appropriate.

Make sure DSpace/dspace-angular#1740 is applied to Angular to allow the identifiers section to be displayed.

Attention to batch imports and other item creation methods by testers is appreciated.

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

kshepherd committed:
[TLC-249] Show Identifier Step, pre-register identifiers

kshepherd committed:
[TLC-249] Remove unused import in WorkspaceItemServiceImpl

kshepherd committed:
[TLC-249] Update section count test in SubmissionDefinitionsControllerIT

tdonohue commented:
@kshepherd : When you find the time, can you resolve the merge conflict in this PR? Thanks!

kshepherd committed:
[TLC-249] Refactor to support better filter usage, PENDING and MINTED status

tdonohue commented:
@kshepherd : Just a note/reminder that this PR looks to have Checkstyle violations and a [few LGTM alerts](https://github.com/DSpace/DSpace/pull/8423#issuecomment-1223413593)...it also looks to be failing some ITs.  When you have a chance to get this updated, I'll gladly give it a closer look...but I'm waiting on the ITs to minimally succeed, as it sounds like this current PR is causing issues in the `PackagerIT`.

kshepherd commented:
@tdonohue thanks, yes I'm still working on a few things to push up and then I'll make sure the ITs and style violations are sorted out - we had some feedback on the use cases and behaviour which caused a bit of additional refactoring to identifier services and DOI providers. The PackagerIT failure is actually a sign that I need to make sure the identiiferService register() method still works as expected when supplying a handle.

kshepherd committed:
[TLC-249] Add and expose config for enabling register DOI

kshepherd committed:
[TLC-249] REST backend for identifier config

kshepherd committed:
[TLC-249] Better status return for new DOIs, write new integration tests

kshepherd commented:
We had the following note from @abollini regarding functionality and use cases for where and now DOIs can be minted or registered. This has inspired some new changes here, though the 'request DOI' submission form button was determined to be out of scope for our particular implementation, the same effect is achievable by using a metadata field with the DOI filter and good autosave settings....

(link to CRIS PR, which I will close soon: https://github.com/4Science/DSpace/pull/252#issuecomment-1210739427)

> Speaking about the functionality: we have implemented something similar as customization for a couple of projects. In our case the request was a bit more complicated because it involve a workflow where the submitter asks to reserve a DOI and someone else need to approve.
The general idea and the implementation that you have provided seems to be a good standard solution to provide out-of-box so we are willing to have it finalized and available in DSpace-CRIS by default (thanks for proposing that!).

>Few remarks: We would prefer to have the minting process triggered manually by the user. The user should click a button to reserve an identifier and not get them reserved automatically all the times (otherwise any test submission will reserve one)
    what we should do if the user reserve a DOI and later on change the item in a way that the DOI should be not longer reserved? we see different options: ignore the filter, the DOI has been reserved so it will be enabled on the item is archived (risky as this could allow user to bypass the filter and could also result in error/low quality metadata sent to datacite); ignore the reservation a message to inform the user that the item doesn't deserve any more the identifier should be provided requiring acknowledgement before proceed; prevent the item to be submitted (validation issue) until the condition to get a DOI are restored
    the scenario where the DOI is reserved but the item is deleted should be tested to verify that orphan reservation create no issue in the minting process

kshepherd committed:
[TLC-249] Integration tests, keep GET handling in controller

kshepherd committed:
[TLC-249] Move GET handling back to ItemIdentifierController

kshepherd committed:
[TLC-249] Checkstyle fixes

kshepherd committed:
[TLC-249] Licence header fixes

kshepherd committed:
[TLC-249] Further checkstyle fixes

kshepherd commented:
@tdonohue there are some tests still failing with this and the angular PR, as well as some possible tidy-up needed to the way I wired up request GET and POST methods for the /identifiers path + link. Some devs from 4Science are helping out by looking at this (both PRs).
I appreciate this is a big change and time is running short for 7.4. If a reschedule is needed to meet deadlines for priority 7.4 release tasks then that is completely understandable.

tdonohue commented:
@kshepherd : Understood.  At this point, because I have so many other reviews on my plate right now, I'm hesitant to review/test this (and the frontend PR) until it's "finalized" (including having passing tests).  So, it's inclusion in 7.4 is becoming less and less likely until it can be finalized.  I'll leave it on the 7.4 board for now, but if it cannot be finalized in the next few days, I suspect we'll have to reschedule it for 7.5...as there just won't be enough time to fully test/review it.

tdonohue commented:
@kshepherd : Moving this to 7.5 as it won't be possible to review/complete in time for 7.4.

kshepherd commented:
> @kshepherd : Moving this to 7.5 as it won't be possible to review/complete in time for 7.4.

Totally understand, thanks for keeping things moving!

kshepherd committed:
[TLC-249] Fix tests, Refactor things to use a LinkRestRepository for GET, Controller for POST

kshepherd committed:
[TLC-249] Further IT fix for identifiers

kshepherd committed:
[TLC-249] Licence header fix

pnbecker commented:
We will resolve the merge conflicts again mid of January. If it is possible to start a review before that, we would highly appreciate that.

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemIdentifierControllerIT.java:
```diff
@@ -0,0 +1,178 @@
+
+        // Expect a successful 201 CREATED for this item with no DOI
+        getClient(token).perform(post("/api/core/items/" +
+                        publicItem1.getID().toString() + "/identifiers"))
```
Maybe I'm reading these tests incorrectly... but is this `POST` sending empty data?  How do we know that means to create a DOI and not a Handle (or "other" identifier)?  The behavior here seems odd, but maybe it's cause I don't have a REST Contract to clarify it.

kshepherd replied:
yes, the POST is simply a "create" operation, and while I was working specifically with DOIs there was no data that could be useful to submit -- in some cases we already have the DOI minted, this is just a manual admin action to say "And now, either mint or promote that DOi to 'to be registered' status". In other cases, we also mint a new DOI first. But the number, prefix, etc., should all come from configuration.

However as per your comments and my answers above, I can see that setting this up for future expansion by at least posting an identifier type, might make things a bit easier to follow.

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemIdentifierControllerIT.java:
```diff
@@ -0,0 +1,178 @@
+        doiService.update(context, doi);
+
+        // Do another POST, again this should return 201 CREATED as we shift the DOI from PENDING to TO_BE_REGISTERED
+        getClient(token).perform(post("/api/core/items/" +
```
Same here, this `POST` seems odd to me at a glance.  Why should an empty data POST result in a DOI moving from one status to the next?  It seems odd that the same endpoint which returns *all* identifiers when doing a `GET` can only perform actions on _DOIs_ when doing a `POST`.

kshepherd replied:
Yep, agreed as above, I can see how this isn't very clear.
DOI happens to be the only identifier type where we can support manual registration by an administrator. And my knowledge of REST and CRUD means POST == "Create" which to me is the kind of operation we're doing.

But we should still POST an identifier type. In the future, we might have a manual "Register ARK" button, and some identifier types might require additional data to be posted.

tdonohue replied:
I'm confused... how is this `POST /api/core/items/[uuid]/identifiers` still a valid command?  I thought this was moved to `POST /api/pid/identifiers`?

I think all the DOI creation tests should move to a new `IdentifierControllerIT` and use the new `/api/pid/identifiers` endpoint.

kshepherd replied:
we've resolvd this as part of the larger refactor

tdonohue commented on dspace/config/modules/identifiers.cfg:
```diff
@@ -0,0 +1,43 @@
+# as the Show Identifiers step which can "pre-mint" DOIs and Handles   #
+#----------------------------------------------------------------------#
+
+# Should DOIs be minted for (future) registration at workspace item creation?
```
I noticed this `identifiers.submission.register=true` also seems to work for Handles.  So, I think we need to update these comments to make it clear this setting is NOT specific to DOIs.

kshepherd replied:
Yes good point, this is one part of the code that is wider than just DOIs.

tdonohue commented on dspace/config/modules/identifiers.cfg:
```diff
@@ -0,0 +1,43 @@
+
+# Show Register DOI button in item status page?
+# Default: false
+#identifiers.item-status.register = true
```
Shouldn't this be named specific to DOIs if it only works for DOIs?  So something like `identifiers.item-status.registerDOI = true`?

kshepherd replied:
yes good call, this should help to make things more clear.

tdonohue reviewed (changes requested):
@kshepherd : Thanks for this work!  Tested this today with https://github.com/DSpace/dspace-angular/pull/1740 and the basics all work.  I can see how this would be of interest to some sites who don't even use DOIs (e.g. if they want to pre-assign Handles).

I also reviewed the code.  Mostly it looks good, but there are a lot of tiny things to clean up (missing/incorrect JavaDocs, etc).  I also noticed the new `/items/[uuid]/identifiers` endpoint is missing a REST Contract, and I had several questions about its design (the `POST` options almost seem too specific to DOIs, with no way to adapt them for other identifiers).

Again, overall, I think this is a good start & it works.  Just needs code cleanup & some discussion around the new `/identifiers` endpoint.

kshepherd commented:
thanks @tdonohue , I'll get onto the tidy-ups and fixes ASAP and think about the best way forward re: abstraction and making some DOI-specific behaviour more obvious. I agree, the empty POST should probably contain an identifier type even if we've only implemented it for DOIs so far.

I was sure i had a new / updated REST contract for the 'create' operation but it might have fallen off somewhere, I'll double check this too

kshepherd commented on dspace/config/item-submission.xml:
```diff
@@ -233,7 +240,7 @@
 
             <step id="collection"/>
-
+            <step id="identifiers"/>
```
i've commented it out with a note above it, referencing the configuration files

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java:
```diff
@@ -112,6 +114,98 @@ public void register(Context context, DSpaceObject dso)
+    @Override
+    public void register(Context context, DSpaceObject dso, Map<Class<? extends Identifier>, Filter> typeFilters)
+            throws AuthorizeException, SQLException, IdentifierException {
+        //We need to commit our context because one of the providers might require the handle created above
```
Yes, obsolete comment, I've removed it now and also tidied surrounding comments

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java:
```diff
@@ -112,6 +114,98 @@ public void register(Context context, DSpaceObject dso)
+    @Override
+    public void register(Context context, DSpaceObject dso, Class<? extends Identifier> type)
+            throws AuthorizeException, SQLException, IdentifierException {
+        //We need to commit our context because one of the providers might require the handle created above
```
Yes, obsolete comment, I've removed it now and also tidied surrounding comments

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java:
```diff
@@ -112,6 +114,98 @@ public void register(Context context, DSpaceObject dso)
+    @Override
+    public void register(Context context, DSpaceObject dso, Class<? extends Identifier> type, Filter filter)
+            throws AuthorizeException, SQLException, IdentifierException {
+        //We need to commit our context because one of the providers might require the handle created above
```
Yes, obsolete comment, I've removed it now and also tidied surrounding comments

kshepherd commented on dspace-api/src/main/java/org/dspace/content/logic/condition/IsArchivedCondition.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    @Override
+    public boolean getResult(Context context, Item item) throws LogicalStatementException {
+        log.debug("Result of isWithdrawn is " + item.isArchived());
```
Thanks, fixed these incorrect references to 'withdrawn'

mwoodiupui replied:
Picky:  the string expression unconditionally allocates a StringBuilder, loads a constant, calls item.isArchived(), converts that to String, calls append() on the StringBuilder, and calls toString() on the StringBuilder, which is a lot of work for a result that will normally be discarded without use.  For future work I'd sugggest a template string and a Producer:

log.debug("Result of isArchived is {}", item::isArchived);

which should simply pass two constants to debug() and generates less garbage.  The Logger is going to look for template tokens whether we use them or not (if it does anything at all).

kshepherd commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/CanRegisterDOIFeature.java:
```diff
@@ -0,0 +1,52 @@
+    @Autowired
+    private AuthorizeServiceRestUtil authorizeServiceRestUtil;
+
+    public static final String NAME = "canRegisterDOI";
```
True. I'll address the DOI vs other identifiers as I can see this is a bit confusing and is an issue in other places you've picked up.

The endpoint is /identiifers, and the idea is a GET will retrieve all the identifiers, and a POST can act as a 'create' - which in the context of our pre-minted DOIs means moving them into "to be registered" status so that on the next DOIOrganiser run, they'll be registered with the provider.

You've noted below that these POSTs are empty, and this is true. I think I should update this to include an identifier type, even if only DOIs are supported right now, to make usage more clear and to help with future support for eg Ark or something.

In this case, the "canRegisterDOI" feature is tied specifically to a button that says "Register DOI". I think we could probably update the name of this feature for the purposes of authZ as it's likely an admin should be able to register any identifier type? But until we start to support other identifier types, I think the button should probably continue to say "Register DOI" just to make it really obvious what happens when you click it (maybe in future it can have a drop down with different identiier types that are valid for manual registration?).

kshepherd committed:
[TLC-249] Address review feedback

Controller updated to be less DOI-specific in its
registration implementation
item-status.register renamed to registerDOI
display types included in section data, for UI hints
comments / javadoc added
DOI status text sent instead of number in status data

As per tdonohue's revew on 2022-01-18

kshepherd committed:
[TLC-249] Lint fixes

kshepherd committed:
[TLC-249] Improve model for identifier step data

kshepherd committed:
[TLC-249] Update configuration identifiers.cfg

kshepherd committed:
[TLC-249] Remove unused subclass DataIdentifier

kshepherd committed:
[TLC-249] Update identifier section IT

tdonohue commented:
@kshepherd : This PR appears to be failing some of its own ITs at this time.  You may already be aware, but wanted to point it out in case you missed it.

pnbecker commented:
@tdonohue @kshepherd is aware of the failing tests. Beside the tests everything else from the reviews should be resolved. So if you have a chance, please take a look again. @kshepherd will work on the tests today.

kshepherd committed:
[TLC-249] Update integration tests for identifier controller

kshepherd committed:
[TLC-249] Fix submission controller IT

kshepherd commented:
@tdonohue I believe I have addressed your concerns here:

 1. As per REST Contract discussion, the data model for the identifiers returned in the identifiers REST endpoint has been made more uniform, and matches the submission section data model for identifiers, too
 2. Simplified display means an identifier will only be displayed in the Submission Section if it exists. There is no longer any placeholder "no DOI has been..." message in either case.
 3. The "Register DOI" button now performs a more generic 'registerIdentifier' operation which takes a required query parameter 'type'. Currently only type=doi is implemented
 4. Unit and integration tests are updated to fix any problems and support this behaviour

corrad82-4s commented on dspace-api/src/main/java/org/dspace/content/logic/FilterUtils.java:
```diff
@@ -0,0 +1,74 @@
+     * Get a map of identifier types and filters to use when creating workspace or archived items
+     * @return
+     */
+    public static Map<Class<? extends Identifier>, Filter> getIdentifierFilters(String status) {
```
Minor: if this method is called only with "workspace" and "install" parameters we could use constants as suggested in method calls, or we can make this method private and expose two distinct public methods one returning "install" (FilterUtils.getIdentifierFilters("install")) and another returning "workspace" filters (FilterUtils.getIdentifierFilters("workspace")).

kshepherd replied:
i've added further javadocs

corrad82-4s commented on dspace-api/src/main/java/org/dspace/ctask/general/RegisterDOI.java:
```diff
@@ -118,8 +119,9 @@ private String register(Item item) {
-            log.debug("Registering DOI with skipFilter = " + skipFilter);
-            doi = provider.register(Curator.curationContext(), item, skipFilter);
+            Filter filter = FilterUtils.getFilterFromConfiguration("identifiers.submission.filter.curation",
+                    new TrueFilter());
```
Minor: could TrueFilter be made singleton?

kshepherd replied:
done

corrad82-4s commented on dspace-api/src/main/java/org/dspace/identifier/FilteredIdentifierProvider.java:
```diff
@@ -24,61 +24,63 @@
-    public void setFilterService(Filter filterService) {
-        this.filterService = filterService;
+    public void setFilter(Filter filter) {
+        this.filter = filter;
```
Minor: since TrueFilter is assigned by default, I would change the code here. I would reassign it if and olny if filter parameter is not null, do something like

Optional.ofNullable(filter).map(f -> this.filter = f)

Behavior should not change

kshepherd replied:
This method did have a null check, but I've updated it to your one-liner as it seems more elegant.

kshepherd replied:
update: this does't compile and I wasn't fully certain of the pattern.
So i replaced it with a simple null check that is better than my original, but also easy to follow and doesn't fail build

corrad82-4s commented on dspace-api/src/main/java/org/dspace/identifier/IdentifierServiceImpl.java:
```diff
@@ -112,11 +114,99 @@ public void register(Context context, DSpaceObject dso)
+    }
+
+    @Override
+    public void register(Context context, DSpaceObject dso, Class<? extends Identifier> type)
```
Any chance to call previous method with null or TrueFilter as "filter" parameter? 
register(Context context, DSpaceObject dso, Class<? extends Identifier> type, Filter filter)

kshepherd replied:
No, I don't think so - I understand the idea here, but note that the `service.register(...)` calls here are actually to a provider (handle provider or DOI provider), not this service.
When the identifier service here doesn't receieve any filter parameter, it should not pass anything onto providers. This allows the providers to use their own defaults without being clobbered (eg. there is a DOI filter specifically set in identifier spring config)

corrad82-4s replied:
Ok, got it. Thank you.

corrad82-4s commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/ItemIdentifierController.java:
```diff
@@ -0,0 +1,166 @@
+        HttpStatus httpStatus = HttpStatus.BAD_REQUEST;
+        ItemRest itemRest;
+        try {
+            DOIIdentifierProvider doiIdentifierProvider = DSpaceServicesFactory.getInstance().getServiceManager()
```
could doiIdentifierProvider be moved as field of this class? If so we could set it during class initialization, possibly raising an exception if no DOI provider is configured

kshepherd replied:
Note: this is now implemented in IdentifierRestRepository.

Wouldn't that cause startup errors for any user that doesn't use DOIs? This REST endpoint needs to be able to start up OK and return data about handles, other identifiers, even if the `createAndReturn` method would throw an error in the case of a missing DOI provider. Does that sound OK?

corrad82-4s replied:
Yes, I see your point and it sounds OK.

corrad82-4s commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:
```diff
@@ -92,6 +93,9 @@ public class ItemRestRepository extends DSpaceObjectRestRepository<Item, ItemRes
     RelationshipTypeService relationshipTypeService;
 
+    @Autowired
+    IdentifierService identifierService;
```
Unless I missed something, this service seems not used anymore

kshepherd replied:
correct, removed

corrad82-4s commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/ShowIdentifiersStep.java:
```diff
@@ -0,0 +1,170 @@
+        try {
+            doi = IdentifierServiceFactory.getInstance().getDOIService().findDOIByDSpaceObject(context, obj.getItem());
+            if (doi != null && !DOIIdentifierProvider.MINTED.equals(doi.getStatus())
+                    && !DOIIdentifierProvider.MINTED.equals(doi.getStatus())) {
```
Minor: not clear to me this third statement

kshepherd replied:
typo - should have excluded both MINTED and DELETED statuses when populating the identifier data section

corrad82-4s reviewed (commented):
Hi, 
I have tested this PR together with its angular part, and all works fine, as expected. 
Overall code looks good to me, just added some (minor) inline considerations.

corrad82-4s commented:
Hi, 
I have another question related to @abollini comment quoted here https://github.com/DSpace/DSpace/pull/8423#issuecomment-1232447108 . 
Unless I missed it (in such case disregard this comment), there's no test verifying 
> the scenario where the DOI is reserved but the item is deleted should be tested to verify that orphan reservation create no issue in the minting process
I would add such test to be sure that there are no issues.

kshepherd committed:
[TLC-249] Update canRegisterDOIFeature to check for cfg enabled

kshepherd committed:
[TLC-249] Refactor identifiers create endpoint

kshepherd commented:
@abollini @corrad82-4s I have pushed my work to refactor endpoints as requested. It is not 100% complete yet and may fail some tests as I haven't updated them yet.

@corrad82-4s I haven't addressed all your feedback regarding spring usage, filters, etc., I will try to get onto this ASAP.

kshepherd committed:
[TLC-249] Refactor, include controller 'find DSO' operation

kshepherd committed:
[TLC-249] Throw not implemented for findOne and findAll

kshepherd commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ItemIdentifierLinkRepository.java:
```diff
@@ -0,0 +1,82 @@
+        try {
+            if (doi != null) {
+                String doiUrl = doiService.DOIToExternalForm(doi.getDoi());
+                IdentifierRest identifierRest = new IdentifierRest(
```
done

kshepherd committed:
[TLC-249] Fixes and improvements as per review feedback

kshepherd commented:
@tdonohue @corrad82-4s @abollini I hope the review requests are all addressed, now. The existing /pid/find endpoint did prove a nuisance when trying to use that namespace, so it is incorporated into my class. I have a feeling this is used by the UI in handle redirects... 

The RestRepository for identifiers does not implement findOne or findAll, and will throw Not Implemented exceptions.
They are not needed for this functionality, and Tomcat's handling of %2F in URL paths makes this really difficult. In the future, if we want full REST operations for identifiers as objects, we should think harder about how to make them addressable consistently.

kshepherd committed:
[TLC-249] Checkstyle fix

kshepherd committed:
[TLC-249] return value  as id for IdentifierRest

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/IdentifierRestRepository.java:
```diff
@@ -0,0 +1,308 @@
+     */
+    @SearchRestMethod(name = "findByItem")
+    @PreAuthorize("permitAll()")
+    public Page<IdentifierRest> findByItem(@Parameter(value = "uuid", required = true)
```
I think we may be missing ITs for `IdentifierRestRepository`?  I'm not seeing an `IdentifierRestRepositoryIT` class.  Probably you should rename the old `IdentifierRestControllerIT` class to be `IdentifierRestRepositoryIT` (since you replaced IdentifierRestController).  Then, I'd recommend adding new tests for:
* findByItem
* creating (minting and queuing)

kshepherd replied:
Yes, I refactored to `IdentifierRestRepositoryIT` which now tests these methods, as well as the pid/find methods used in the legacy controller

tdonohue commented on dspace/config/spring/api/item-filters.xml:
```diff
@@ -335,5 +346,27 @@
         </property>
     </bean>
 
+    <bean id="example-doi_filter" class="org.dspace.content.logic.DefaultFilter">
```
Could we add a basic comment to this to describe what it might be used for?  Right now, it's here as an "example", but it's unclear (to me at least) why DOIs might want to use such a filter.  It'd be good to describe this a bit more in a code comment or similar.

(Or if this example isn't needed anymore, then we should remove it.  Just saying it doesn't make much sense right now without some sort of explanation)

kshepherd replied:
Good catch, I actually removed it in the end.

tdonohue reviewed (commented):
@kshepherd : Gave this a quick code review today (haven't had a chance to re-test, but hoping to do so later today/tomorrow).  It's looking better to me overall.  I think it's missing some tests though (see inline comments) and/or some old tests need to be migrated to using the new /api/pid/identifiers endpoint.  I think this is also failing the CI build cause of minor checkstyle issues.

Overall, it's looking like a good direction to me.  Just more minor cleanup needed.

(BTW, I think it's perfectly OK to not implement findOne or findAll for these endpoints.  We can work on a better way to achieve those once they are actually necessary.  Your current implementation here looks fine to me.)

tdonohue commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/submit/step/ShowIdentifiersStep.java:
```diff
@@ -0,0 +1,170 @@
+        }
+
+        // Populate bean with data and return, if the identifier type is configured for exposure
+        result.addIdentifier("doi", doiString,
```
Is the intention here to always have a "doi" shown in this step *even if* the DOI is empty?  I noticed that this results in the JSON always including a DOI section even if it's empty.  For instance, if DOIs are disabled but I'm pre-registering Handles, I'm seeing an identifier section like this:
```
    "identifiers" : [ {
        "id" : null,
        "value" : null,
        "identifierType" : "doi",
        "identifierStatus" : null,
        "type" : "identifier"
      }, {
        "id" : "http://localhost:4000/handle/123456789/7183",
        "value" : "http://localhost:4000/handle/123456789/7183",
        "identifierType" : "handle",
        "identifierStatus" : null,
        "type" : "identifier"
      } ],
      "displayTypes" : [ "handle" ]
```

It just seems odd to me that we'd include a "doi" or a "handle" when they are *null*.  I'd expect us to have a `isNotEmpty()" check around both of these `.addIdentifier()` calls.

That said, this is really minor... if it's needed, we can leave it.  But, it just looks odd to me that we have null DOIs or Handles returned in this step.

kshepherd replied:
@tdonohue in this case, my intention was to establish a difference between "not getting a DOI" (null) and non-existant (not configured / no DOIs used)- it is so the section can explicitly say "i'm configured to display DOIs, but one will *not* be minted for this item as things currently stand".

tdonohue replied:
@kshepherd : I guess I was just surprised that *with DOIs disabled* (as is in my setup), I still see a `null` DOI listed.  I can understand though how this would be useful when DOIs are *enabled*.

That said, as I mentioned, this is super minor.  If it's better to leave this as-is, it's fine with me.  It just struck me as "odd" since I don't even have DOIs enabled

kshepherd replied:
yes, fair enough. i think there are several areas which can be improved in this already-large feature, i'll expand on that a bit in my final comment after pushing changes, but i hope that these "areas for improvement" won't be blockers -- we'll see!

tdonohue replied:
Nope, not a blocker. As I said, it's "odd", but it'd only be a blocker if it was *assigning* a DOI :)  It's just a random null DOI, which we could clean up later if we need to.  If there's an easy fix though, feel free.

kshepherd commented:
@tdonohue i'm working on this right now so you shoudl see my ITs renamed and updated appropriately to match the recent refactors, I'll leave another message with notes for the reviewers once it is ready again

kshepherd committed:
[TLC-249] Fix category type in IdentifierRest

kshepherd committed:
[TLC-249] Update integration tests for IdentifierRestRepository

kshepherd committed:
[TLC-249] Refactor TrueFilter to get instantiated as singleton

kshepherd committed:
[TLC-249] Construct canonical handle URL in identifier item link

kshepherd committed:
[TL-249] Remove example-doi_filter from config, add singleton attr to TrueFilter

kshepherd committed:
[TL-249] Add always_true_filter to test REST identifier-service.xml config

kshepherd committed:
[TL-249] Add identifiers.cfg and item-filters.xml config to test data

kshepherd committed:
[TL-249] Fix imports

kshepherd committed:
[TL-249] Revert accidental commit to RestResourceController

kshepherd commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/RestResourceController.java:
```diff
@@ -125,6 +125,7 @@ public void afterPropertiesSet() {
             // this doesn't work as we don't have an active http request
             // see https://github.com/spring-projects/spring-hateoas/issues/408
             // Link l = linkTo(this.getClass(), r).withRel(r);
+            log.error(r);
```
Whoops! Corrected

kshepherd committed:
[TL-249] Javadoc and other small fixes, DOI unit tests

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/DOIIdentifierProvider.java:
```diff
@@ -471,17 +471,22 @@ public void updateMetadata(Context context, DSpaceObject dso, String identifier)
-        DOI doiRow = loadOrCreateDOI(context, dso, doi, skipFilter);
+        DOI doiRow = loadOrCreateDOI(context, dso, doi, updateFilter);
+
+        if (PENDING.equals(doiRow.getStatus())) {
```
i've added tests in `DOIIdentifierProviderTest` which should satisfy this

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/service/IdentifierService.java:
```diff
@@ -103,6 +105,40 @@ void reserve(Context context, DSpaceObject dso, String identifier)
+     * @throws SQLException        if database error
+     * @throws IdentifierException if identifier error
+     */
+    void register(Context context, DSpaceObject dso, Class<? extends Identifier> type)
```
done

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/service/IdentifierService.java:
```diff
@@ -103,6 +105,40 @@ void reserve(Context context, DSpaceObject dso, String identifier)
+     * @throws SQLException        if database error
+     * @throws IdentifierException if identifier error
+     */
+    void register(Context context, DSpaceObject dso, Class<? extends Identifier> type, Filter filter)
```
done

kshepherd commented on dspace-api/src/main/java/org/dspace/identifier/service/IdentifierService.java:
```diff
@@ -103,6 +105,40 @@ void reserve(Context context, DSpaceObject dso, String identifier)
+     * @throws SQLException        if database error
+     * @throws IdentifierException if identifier error
+     */
+    void register(Context context, DSpaceObject dso, Map<Class<? extends Identifier>, Filter> typeFilters)
```
done

kshepherd commented on dspace-api/src/main/java/org/dspace/content/logic/condition/IsArchivedCondition.java:
```diff
@@ -0,0 +1,38 @@
+    private final static Logger log = LogManager.getLogger();
+
+    /**
+     * Return true if item is withdrawn
```
fixed

kshepherd commented on dspace-api/src/main/java/org/dspace/content/WorkspaceItemServiceImpl.java:
```diff
@@ -160,6 +171,26 @@ public WorkspaceItem create(Context context, Collection collection, UUID uuid, b
+                .getBooleanProperty("identifiers.submission.register", false)) {
+            try {
+                // Get map of filters to use for identifier types
+                Map<Class<? extends Identifier>, Filter> filters = FilterUtils.getIdentifierFilters("workspace");
```
Yes, good call, I actually refactored even more simply. to a boolean `inProgress` to simply indicate whether the item is in progress or not. The strings are used just in that specific case for building a configuration property key

kshepherd commented on dspace-api/src/main/java/org/dspace/content/logic/FilterUtils.java:
```diff
@@ -0,0 +1,74 @@
+     * @param property  DSpace configuration property name (Apache Commons config)
+     * @return  Filter or null
+     */
+    public static Filter getFilterFromConfiguration(String property, Filter defaultFilter) {
```
Yes, I've added better javadoc to make this clear. The value of the config property is the bean ID.

kshepherd commented on dspace-api/src/main/java/org/dspace/content/logic/FilterUtils.java:
```diff
@@ -0,0 +1,74 @@
+     * @param property  DSpace configuration property name (Apache Commons config)
+     * @return  Filter or null
+     */
+    public static Filter getFilterFromConfiguration(String property) {
```
Yes, I've added better javadoc to make this clear. The value of the config property is the bean ID.

kshepherd commented on dspace-api/src/main/java/org/dspace/content/logic/Filter.java:
```diff
@@ -32,4 +33,6 @@ public interface Filter extends LogicalStatement {
     @Override
     boolean getResult(Context context, Item item) throws LogicalStatementException;
+
+    String getName();
```
done

kshepherd committed:
[TL-249] Javadoc and other small fixes

Including stripping $Revision$ from filter javadoc.

kshepherd committed:
[TL-249] Checkstyle fix

kshepherd commented:
@tdonohue @corrad82-4s 

Thanks for your feedback on this PR!

I have addressed every comment, along with implementing the refactor of endpoint usage outlined in the RestContract PR.

Sorry for the length of this PR! The comments are a little hard to wade through now (most of our conversations are hidden behind UI accordions)

I think there are improvements that can be made here: I would prefer to use a paginated list in the item/{uuid}/identifiers link to be identical to the findByItem method (ie. just redirect to that method), but given the size of the refactor and the time constraints, I ended up with a slightly different model used for that specific purpose (display of DOI in status page), though it is extremely close and works well. I would look to address this in a future release as it would require service and model changes in angular.

If any reviewers are running practical tests, as well as the DOI testing Tim mentioned, some lesser-known ways of importing / ingesting content would be great to test too.

corrad82-4s commented:
Hi @kshepherd , 
thank you for addressing my feedbacks. I saw that also what discussed on RestContract PR (https://github.com/DSpace/RestContract/pull/199#pullrequestreview-1274846262) is addressed. 

Unless I missed it, are we verifying and how that orphan reservations are not creating issues with minting process? (https://github.com/DSpace/DSpace/pull/8423#issuecomment-1406867052)

pnbecker committed:
[TLC-249] Test abandoning of a pending DOI and minting afterwards

pnbecker commented:
@corrad82-4s I added the missing test.

pnbecker committed:
[TLC-249] Ensure property identifiers.item-status.register-doi is set

pnbecker commented:
@tdonohue I retested it with DOIs. @kshepherd and me noticed that a configuration property exposed via rest and used by angular did not had a default value. I changed this. There is one last discussion going on in the angular part of this PR. Besides that this PR is ready.

corrad82-4s reviewed (approved):
Feedbacks have been addressed, as well as required tests.

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  Looks good now
